File: ui/frontend/app/models/item.js
import Model, { attr, belongsTo } from '@ember-data/model';

export default class ItemModel extends Model {
  @attr('number', { defaultValue: 0 }) notificationCount;
  @attr('number') price;
  @attr('number', { defaultValue: 0 }) descendantsOwe;
  @attr('number') payUp;
  @attr('boolean') isPaid;

  @belongsTo() invoice;
  @belongsTo() account;
}

File: backend/app/models/item.rb
class Item < ApplicationRecord
  after_commit :check_price_updated 
  
  belongs_to :account
  belongs_to :invoice
  
  def pay_up
    descendants_owe + price
  end
  
  before_save { |item|
    if (item.notification_count_was != item.notification_count) and (item.account.email)
      ItemMailer.invoice_email(item).deliver_now
    end
  } 
  
  def check_price_updated
    if self.saved_change_to_attribute? :price
      self.recalculate_descendants_owe
    end
  end
  
  def recalculate_descendants_owe
    # This one might be tricky, better wrap in transaction
    invoice = self.invoice
    invoice.with_lock do
      ancestor_accounts = self.account.self_and_ancestors
      ancestor_items = Item.where(invoice: invoice, account: ancestor_accounts)
      ancestor_items.each do |item|
        descendant_accounts = item.account.descendants
        sum = Item.where(invoice: invoice, account: descendant_accounts).sum(:price)
        item.descendants_owe = sum
        item.save!
      end
    end
  end
end
